### ISSUE TITLE
`unnecessary_lazy_evaluations` suggests wrong fix

### ISSUE THREAD

**Issue by barafael** [rust-lang/rust-clippy#11672]
### Summary

`unnecessary_lazy_evaluations` is normally a great lint, but in this example it produces a fix that does not compile.

The realistic example comes from [embassy/embassy-usb](https://github.com/embassy-rs/embassy/blob/cd92bc3145f1cc7e0182a45dd898bc737e5587b6/embassy-usb/src/lib.rs#L388-L390).

### Reproducer

I tried this code ([see playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn+main%28%29+%7B%0A++++let+data+%3D+vec%21%5B1%2C+2%2C+3%5D%3B%0A++++let+len+%3D+3%3B%0A++++let+max_packet_size+%3D+2%3B%0A++++let+need_zlp+%3D+true%3B%0A++++let+chunks+%3D+data%5B0..len%5D%0A++++++++.chunks%28max_packet_size%29%0A++++++++.chain%28need_zlp.then%28%7C%7C+-%3E+%26%5Bu8%5D+%7B+%26%5B%5D+%7D%29%29%3B%0A++++++++%2F%2F.chain%28need_zlp.then_some%28%7B+%26%5B%5D+%7D%29%29%3B%0A++++dbg%21%28chunks%29%3B%0A%7D%0A)):

```rust
fn main() {
    let data = vec![1, 2, 3];
    let len = 3;
    let max_packet_size = 2;
    let need_zlp = true;
    let chunks = data[0..len]
        .chunks(max_packet_size)
        .chain(need_zlp.then(|| -> &[u8] { &[] }));
        //.chain(need_zlp.then_some({ &[] }));
    dbg!(chunks);
}
```

I expected to see this happen: No lint or a valid correction

Instead, this happened:

```rust
warning: unnecessary closure used with `bool::then`
 --> src/main.rs:8:16
  |
8 |         .chain(need_zlp.then(|| -> &[u8] { &[] }));
  |                ^^^^^^^^^-------------------------
  |                         |
  |                         help: use `then_some(..)` instead: `then_some({ &[] })`
```

### Version

```text
rustc 1.75.0-nightly (e0d7ed1f4 2023-10-01)
binary: rustc
commit-hash: e0d7ed1f453fb54578cc96dfea859b0e7be15016
commit-date: 2023-10-01
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Jarcho** [comment]
This wasn't actually fixed, it just won't be auto-fixed.
